### ISSUE TITLE
Object initialization with dependency via case class deadlocks 

### ISSUE THREAD

**Issue by laurynaslubys** [lampepfl/dotty#7960]
## minimized code

```Scala
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._
import scala.concurrent.{Await, Future}

class A() {
  val bc = B.C(50)
}

object A {
  Thread.sleep(100L)

  val a = new A()
}


object B {
  Thread.sleep(100L)

  case class C(a: Int)

  val a: A = A.a
}

object Repro {
  final def main(args: Array[String]): Unit = {
    Await.result(Future.sequence(Seq(
      Future { A.a },
      Future { B.a },
    )), 1.seconds)
  }
}

```

<details>
<summary>Compilation output</summary>

```
[error] Exception in thread "main" java.util.concurrent.TimeoutException: Future timed out after [1 second]
```
</details>

## expectation
I would expect initialization to not deadlock. This snippet does not deadlock in scala 2.x. 

The `sleep`s are there just to make the issue more easily reproducible, without it the issue is still present, but intermittent.

I see no reason for the initialization to deadlock: initializing an instance of `A` does not require `object B` to be initialized as `case class C` does not reference any fields on `object B`.

**Comment by griggt** [comment]
This may have been fixed by d65c0d9 in #9767.  That is to say, I have been unable to get it to repro from that commit forward.

**Comment by smarter** [comment]
Seems likely, do you mind making a PR adding this testcase as a run test?
